File: jsrc/legacy/SxxMachine/builtin/PRED_mutex_lock_bt_1.java
package com.googlecode.prolog_cafe.builtin;

import com.googlecode.prolog_cafe.lang.*;
import com.googlecode.prolog_cafe.lang.Predicate.P1;

import java.util.concurrent.locks.Lock;
/**

<p>mutex_lock(+MutexId)
    <p>Lock the mutex.   Prolog mutexes are recursive mutexes:  they can be
    locked  multiple times by the same thread.  Only after  unlocking it
    as  many times as it is  locked does the mutex become available  for
    locking  by other threads.   If another thread has locked the  mutex
    the calling thread is suspended until the mutex is unlocked.

    <p>If  MutexId is  an atom,  and there  is no current  mutex with  that
    name,  the  mutex  is  created automatically  using  mutex_create/1.
    This implies named mutexes need not be declared explicitly.

    <p>Please  note that  locking and  unlocking mutexes  should be  paired
    carefully.     Especially  make  sure  to  unlock  mutexes  even  if
    the  protected  code  fails  or  raises an  exception.     For  most
    common  cases,  use with_mutex/2,  which provides  a  safer way  for
    handling  Prolog-level mutexes.   The predicate setup_call_cleanup/3
    is  another  way  to guarantee  that  the  mutex is  unlocked  while
    retaining non-determinism.

    <p>the returned mutex is automatically unlocked on backtracking 

  <p>description is copied from swi prolog documentation
 * @author semenov
 *
 */
public class PRED_mutex_lock_bt_1 extends P1 {

	public PRED_mutex_lock_bt_1(Term arg1, Operation cont) {
		this.arg1 = arg1;
		this.cont = cont;
	}

	@Override
	public Operation exec(Prolog engine) throws PrologException {
		Term a1 = arg1.dereference();

		final Lock lock;
		if ((a1 instanceof SymbolTerm)){
			lock = Mutex.getInstance(a1.name());
		} else if ((a1 instanceof JavaObjectTerm) && (a1.toJava() instanceof Lock)) {
			lock = (Lock) a1.toJava();
		} else {
			throw new IllegalTypeException(this, 1, "atom or JavaObjectTerm(Lock)", a1);
		}
		lock.lock();
		engine.trail.push(() -> {
			try {
				lock.unlock();
			} catch (IllegalMonitorStateException e){
				// ignore, because the lock is already unlocked explicitly
			}
		});
		return cont;
	}
}

File: src/legacy/SxxMachine/builtin/PRED_mutex_lock_bt_1.cpp
using namespace std;

#include "PRED_mutex_lock_bt_1.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../exceptions/SxxMachine/PrologException.h"
#include "../../../machine/SxxMachine/SymbolTerm.h"
#include "../../../machine/SxxMachine/Mutex.h"
#include "../../../machine/SxxMachine/FFIObjectTerm.h"
#include "../../../exceptions/SxxMachine/IllegalTypeException.h"

namespace SxxMachine
{
	using IllegalTypeException = SxxMachine::IllegalTypeException;
	using PrologException = SxxMachine::PrologException;
	using FFIObjectTerm = SxxMachine::FFIObjectTerm;
	using Mutex = SxxMachine::Mutex;
	using Operation = SxxMachine::Operation;
	using P1 = SxxMachine::Predicate::P1;
	using Prolog = SxxMachine::Prolog;
	using SymbolTerm = SxxMachine::SymbolTerm;
	using Term = SxxMachine::Term;

	PRED_mutex_lock_bt_1::PRED_mutex_lock_bt_1(Term *a0, Operation cont)
	{
		this->LARG[0] = a0; //0];
		this->cont = cont;
	}

	Operation PRED_mutex_lock_bt_1::exec(Prolog *engine) throw(PrologException)
	{
		Term *a1 = LARG[0]->dref();

		Lock * const lock;
		if ((dynamic_cast<SymbolTerm*>(a1) != nullptr))
		{
			lock = Mutex::getInstance(a1->name());
		}
		else if ((dynamic_cast<FFIObjectTerm*>(a1) != nullptr) && (dynamic_cast<Lock*>(a1->toJava()) != nullptr))
		{
			lock = any_cast<Lock*>(a1->toJava());
		}
		else
		{
			throw IllegalTypeException(this, 1, "atom or FFIObjectTerm(Lock)", a1);
		}
		lock->lock();
		engine->trail->push([&] ()
		{
		try
		{
			lock->unlock();
		}
		catch (const IllegalMonitorStateException &e)
		{
		}
		});
		return cont;
	}
}
